src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/GVMDependenciesNode.cs

```diff
@@ -0,0 +1,178 @@
+                EETypeNode entryAsEETypeNode = entry as EETypeNode;
+
+                if (entryAsEETypeNode == null)
+                    continue;
+
+                TypeDesc potentialOverrideType = entryAsEETypeNode.Type;
+                if (!(potentialOverrideType is DefType))
+                    continue;
+
+                Debug.Assert(!potentialOverrideType.IsRuntimeDeterminedSubtype);
+
+                if (_method.OwningType.HasSameTypeDefinition(potentialOverrideType) && potentialOverrideType.IsInterface && (potentialOverrideType != _method.OwningType))
+                {
+                    if (_method.OwningType.CanCastTo(potentialOverrideType))
+                    {
+                        // Variance expansion
+                        MethodDesc matchingMethodOnRelatedVariantMethod = potentialOverrideType.GetMethod(_method.Name, _method.GetTypicalMethodDefinition().Signature);
+                        matchingMethodOnRelatedVariantMethod = _method.Context.GetInstantiatedMethod(matchingMethodOnRelatedVariantMethod, _method.Instantiation);
+                        dynamicDependencies.Add(new CombinedDependencyListEntry(factory.GVMDependencies(matchingMethodOnRelatedVariantMethod), null, "GVM Variant Interface dependency"));
+                    }
+                }
+
+                // If this is an interface gvm, look for types that implement the interface
+                // and other instantantiations that have the same canonical form.
+                // This ensure the various slot numbers remain equivalent across all types where there is an equivalence
+                // relationship in the vtable.
+                if (_method.OwningType.IsInterface)
+                {
+                    if (potentialOverrideType.IsInterface)
+                        continue;
+
+                    foreach (DefType interfaceImpl in potentialOverrideType.RuntimeInterfaces)
+                    {
+                        if (interfaceImpl.ConvertToCanonForm(CanonicalFormKind.Specific) == _method.OwningType.ConvertToCanonForm(CanonicalFormKind.Specific))
+                        {
+                            // Find if the type implements this method. (Note, do this comparision against the generic definition of the method, not the
+                            // specific method instantiation that is "method"
+                            MethodDesc genericDefinition = interfaceImpl.GetMethod(_method.Name, _method.GetTypicalMethodDefinition().Signature);
+
+                            MethodDesc slotDecl;
+                            TypeDesc currentType = potentialOverrideType;
+                            do
+                            {
+                                slotDecl = currentType.ResolveInterfaceMethodToVirtualMethodOnType(genericDefinition);
+                                currentType = currentType.BaseType;
+                            }
+                            while (slotDecl == null && currentType != null);
```

Comment:
This is a very odd algorithm. As such we should make it testable. Could you put it into the typesystem dll as an extension method or something, and then write some tests?